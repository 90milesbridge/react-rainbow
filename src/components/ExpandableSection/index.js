/* eslint-disable react/require-default-props */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import IconSvg from '../IconSvg';
import { uniqueId } from '../../libs/utils';

export default class ExpandableSection extends Component {
    constructor(props) {
        super(props);
        this.state = {
            isExpanded: this.props.initiallyExpanded,
            id: uniqueId('expandable-section'),
        };

        this.handleExpanding = this.handleExpanding.bind(this);
    }

    getContainerClass() {
        return classnames('slds-section', {
            'slds-is-open': this.state.isExpanded,
        }, this.props.className);
    }

    handleExpanding() {
        const { onExpandOrCollapse } = this.props;
        const nextState = !this.state.isExpanded;

        this.setState({ isExpanded: nextState });
        onExpandOrCollapse(this.state);
    }

    renderHeader() {
        const { label } = this.props;
        const { id, isExpanded } = this.state;

        return (
            <h3 className="slds-section__title">
                <button aria-controls={id} aria-expanded={isExpanded} className="slds-button slds-section__title-action" onClick={this.handleExpanding}>
                    <IconSvg className="slds-section__title-action-icon slds-button__icon slds-button__icon_left" iconName="utility:switch" variant="bare" />
                    {label ? <span className="slds-truncate">{label}</span> : null}
                </button>
            </h3>
        );
    }

    render() {
        const { children, style } = this.props;

        return (
            <div className={this.getContainerClass()} style={style}>
                {this.renderHeader()}
                <div aria-hidden={!this.state.isExpanded} className="slds-section__content" id={this.state.id}>
                    {children}
                </div>
            </div>
        );
    }
}

ExpandableSection.propTypes = {
    /** A CSS class for the outer element, in addition to the component's base classes. */
    className: PropTypes.string,
    /** An object with custom style applied to the outer element. */
    style: PropTypes.object,
    /** It tells if the section is expanded initially. This value only change the
     state of the section one time */
    initiallyExpanded: PropTypes.bool,
    /** The title in header of the component */
    label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
    /** The content body section */
    children: PropTypes.node,
    /** The callback is fired when the section expand or collapse. Its invoked with
     the current state of the section like { isExpanded: true } */
    onExpandOrCollapse: PropTypes.func,
};

ExpandableSection.defaultProps = {
    initiallyExpanded: false,
    onExpandOrCollapse: () => {},
};
