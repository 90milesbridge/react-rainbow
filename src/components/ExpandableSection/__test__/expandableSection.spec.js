import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import ExpandableSection from './../index';

describe('<ExpandableSection/>', () => {
    it('should the right markup with the right classes', () => {
        const component = renderer.create(
            <ExpandableSection label="Section Title" initiallyExpanded >
                Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum.
                Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh, ut
                fermentum massa justo sit amet risus. Lorem ipsum dolor sit amet, consectetur
                adipiscing elit. Nullam quis risus eget urna mollis ornare vel eu leo. Nulla
                vitae elit libero, a pharetra augue.
            </ExpandableSection>,
        );

        expect(component).toMatchSnapshot();
    });

    it('should update aria-expanded attribute when expand/collapse', () => {
        const component = shallow(<ExpandableSection label="Header">Some text</ExpandableSection>);
        const button = component.find('button');

        expect(component.find('button[aria-expanded=false]').exists()).toBe(true);
        button.simulate('click');
        expect(component.find('button[aria-expanded=true]').exists()).toBe(true);
    });

    it('should expand when we click on the header and it was collapse', () => {
        const component = shallow(<ExpandableSection label="Header">Some text</ExpandableSection>);
        const button = component.find('button');

        expect(component.find('div.slds-is-open').exists()).toBe(false);
        button.simulate('click');
        expect(component.find('div.slds-is-open').exists()).toBe(true);
    });

    it('should collapse when we click on the header and it was expanded', () => {
        const component = shallow(<ExpandableSection initiallyExpanded label="Header">Some text</ExpandableSection>);
        const button = component.find('button');

        expect(component.find('div.slds-is-open').exists()).toBe(true);
        button.simulate('click');
        expect(component.find('div.slds-is-open').exists()).toBe(false);
    });

    it('should fire onExpandOrCollapse function when the state of the section change', () => {
        const onExpandOrCollapseMockFn = jest.fn();
        const component = shallow(
            <ExpandableSection onExpandOrCollapse={onExpandOrCollapseMockFn} label="Header">
                Some text
            </ExpandableSection>,
        );
        const button = component.find('button');

        button.simulate('click');
        expect(onExpandOrCollapseMockFn.mock.calls.length).toBe(1);
    });

    it('should pass the current state of the section when onExpandOrCollapse function is invoked', () => {
        const callbackMockFn = jest.fn();
        const component = shallow(
            <ExpandableSection onExpandOrCollapse={callbackMockFn} label="Header">
                Some text
            </ExpandableSection>,
        );
        const button = component.find('button');

        button.simulate('click');
        expect(callbackMockFn.mock.calls[0][0].isExpanded).toBe(false);
    });
});
