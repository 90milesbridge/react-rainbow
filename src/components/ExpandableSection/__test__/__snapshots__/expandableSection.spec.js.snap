// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExpandableSection/> should the right markup with the right classes 1`] = `
<div
  className="slds-section slds-is-open"
  style={undefined}
>
  <h3
    className="slds-section__title"
  >
    <button
      aria-controls="expandable-section-1"
      aria-expanded={true}
      className="slds-button slds-section__title-action"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="slds-section__title-action-icon slds-button__icon slds-button__icon_left"
        style={undefined}
      >
        <use
          xlinkHref="/icons/utility-sprite/svg/symbols.svg#switch"
          xmlnsXlink="http://www.w3.org/1999/xlink"
        />
      </svg>
      <span
        className="slds-truncate"
      >
        Section Title
      </span>
    </button>
  </h3>
  <div
    aria-hidden={false}
    className="slds-section__content"
    id="expandable-section-1"
  >
    Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum. Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh, ut fermentum massa justo sit amet risus. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam quis risus eget urna mollis ornare vel eu leo. Nulla vitae elit libero, a pharetra augue.
  </div>
</div>
`;
